Core/GDCore/Project/ObjectFolderOrObject.cpp

```diff
@@ -0,0 +1,241 @@
+
+using namespace std;
+
+namespace gd {
+
+ObjectFolderOrObject ObjectFolderOrObject::badObjectFolderOrObject;
+
+ObjectFolderOrObject::ObjectFolderOrObject() : folderName("__NULL") {}
+ObjectFolderOrObject::ObjectFolderOrObject(gd::String folderName_,
+                                           ObjectFolderOrObject* parent_)
+    : folderName(folderName_), parent(parent_) {}
+ObjectFolderOrObject::ObjectFolderOrObject(gd::Object* object_,
+                                           ObjectFolderOrObject* parent_)
+    : object(object_), parent(parent_) {}
+ObjectFolderOrObject::~ObjectFolderOrObject() {}
+
+bool ObjectFolderOrObject::HasObjectNamed(const gd::String& name) {
+  if (IsFolder()) {
+    return std::any_of(
+        children.begin(),
+        children.end(),
+        [&name](
+            std::unique_ptr<gd::ObjectFolderOrObject>& objectFolderOrObject) {
+          return objectFolderOrObject.get()->HasObjectNamed(name);
+        });
+  }
+  return object->GetName() == name;
+}
+ObjectFolderOrObject& ObjectFolderOrObject::GetObjectNamed(
+    const gd::String& name) {
+  if (!IsFolder() && object->GetName() == name) {
+    return *this;
+  }
+  if (IsFolder()) {
+    for (std::size_t j = 0; j < children.size(); j++) {
+      ObjectFolderOrObject& foundInChild = children[j]->GetObjectNamed(name);
+      if (&(foundInChild) != &badObjectFolderOrObject) {
+        return foundInChild;
+      }
+    }
+  }
+  return badObjectFolderOrObject;
+}
+
+void ObjectFolderOrObject::SetFolderName(const gd::String& name) {
+  if (!IsFolder()) return;
+  folderName = name;
+}
+
+ObjectFolderOrObject& ObjectFolderOrObject::GetChildAt(std::size_t index) {
+  return *children[index];
+}
+ObjectFolderOrObject& ObjectFolderOrObject::GetObjectChild(
+    const gd::String& name) {
+  for (std::size_t j = 0; j < children.size(); j++) {
+    if (!children[j]->IsFolder()) {
+      if (children[j]->GetObject().GetName() == name) return *children[j];
+    };
+  }
+  return badObjectFolderOrObject;
+}
+
+void ObjectFolderOrObject::InsertObject(gd::Object* insertedObject) {
+  auto objectFolderOrObject =
+      gd::make_unique<ObjectFolderOrObject>(insertedObject, this);
+  children.push_back(std::move(objectFolderOrObject));
+}
+void ObjectFolderOrObject::InsertObject(gd::Object* insertedObject,
+                                        std::size_t position) {
+  auto objectFolderOrObject =
+      gd::make_unique<ObjectFolderOrObject>(insertedObject, this);
+  children.insert(
+      position < children.size() ? children.begin() + position : children.end(),
+      std::move(objectFolderOrObject));
+}
+
+std::size_t ObjectFolderOrObject::GetChildPosition(
+    ObjectFolderOrObject& child) const {
+  for (std::size_t j = 0; j < children.size(); j++) {
+    if (children[j].get() == &child) return j;
+  }
+  return gd::String::npos;
+}
+
+ObjectFolderOrObject& ObjectFolderOrObject::InsertNewFolder(
+    const gd::String newFolderName, std::size_t position) {
```

Comment:
```suggestion
    const gd::String& newFolderName, std::size_t position) {
```

(in 99% of case, you can pass a const ref to a string, even from the JS)